### DISCUSSION TITLE
Get container ID while using CDI (Container Device Interface)

### DISCUSSION THREAD

**Discussion by cjvolzka** [moby/moby#48799] [Q&A] [answered]
Is it possible to get the container ID (or container name) being created using CDI or the container start hooks?

Background: I'd like to use CDI (https://github.com/moby/moby/pull/45134 and https://github.com/docker/cli/pull/4084) to add devices to a container. The device should not be sharable between containers but it appears CDI does not block multiple containers from requesting the same device. 

I'm exploring using the hooks part of the CDI spec to prevent this. Basically an application (called via the hook) will track which devices are in use and fail if it sees another container is already using the device.

I'd like the application to know which device was requested (can already get this from CDI) and which container ID requested it. However I'm unsure how or where I'd get the container ID from the hooks {createRuntime, createContainer, startContainer} to pass it to the application. Is there a way accessible by an application called via a hook to get that (env variable, file, psuedo file, or process)?

**Comment by thaJeztah** [comment, marked as answer]
I'm not a full expert on everything CDI, but when OCI hooks are invoked (see the [OCI runtime specification](https://github.com/opencontainers/runtime-spec/blob/v1.2.0/config.md#posix-platform-hooks)), the binary that's invoked will get the container's State passed through STDIN;

> Hooks allow users to specify programs to run before or after various lifecycle events. Hooks MUST be called in the listed order. The state of the container MUST be passed to hooks over stdin so that they may do work appropriate to the current state of the container.

One of the fields of the State struct that's passed is `ID`, which is the container's ID; https://github.com/opencontainers/runtime-spec/blob/v1.2.0/runtime.md#state

```json
{
    "ociVersion": "0.2.0",
    "id": "oci-container1",
    "status": "running",
    "pid": 4422,
    "bundle": "/containers/redis",
    "annotations": {
        "myKey": "myValue"
    }
}
```

**Reply by cjvolzka** [reply]
Thanks @thaJeztah! This is exactly what I was looking for. Bonus points for including the spec link!

**Reply by thaJeztah** [reply]
Perfect! Hope you get it all wired up 👍
